### PR TITLE
Replace FieldStatsProvider with a method on MappedFieldType.

### PR DESCRIPTION
**Pull request by jpountz**
FieldStatsProvider had to perform instanceof calls to properly handle dates or
ip addresses. By moving the logic to MappedFieldType, each field type can check
whether all values are within bounds its way.

Note that this commit only keeps rewriting support for dates, which are the only
field for which the rewriting mechanism is likely to help (because of time-based
indices).

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/index/mapper/core/DateFieldTypeTests.java [hunk 1]
```diff
@@ -54,4 +67,49 @@ public void modify(MappedFieldType ft) {
...
+
+    public void testIsFieldWithinQueryEmptyReader() throws IOException {
+        IndexReader reader = new MultiReader();
+        DateFieldType ft = new DateFieldType();
+        ft.setName("my_date");
+        assertEquals(Relation.DISJOINT, ft.isFieldWithinQuery(reader, "2015-10-12", "2016-04-03",
+                randomBoolean(), randomBoolean(), null, null));
+    }
+
+    public void testIsFieldWithinQuery() throws IOException {
```
Comment (colings86) [inline core/src/test/java/org/elasticsearch/index/mapper/core/DateFieldTypeTests.java]: Should we add a test to each of the other types to check that for random query bounds `MappedFieldType.isFieldWithinQuery(IndexReader, Object, Object, boolean, boolean, TimeZone, DateMathParser)` always  returns Relation.INTERSECTS?

File: core/src/test/java/org/elasticsearch/index/mapper/core/DateFieldTypeTests.java [hunk 2]
```diff
@@ -54,4 +67,49 @@ public void modify(MappedFieldType ft) {
...
+        assertEquals(Relation.WITHIN, ft.isFieldWithinQuery(reader, "2015-10-12", "2016-04-03",
+                true, true, null, null));
+        assertEquals(Relation.INTERSECTS, ft.isFieldWithinQuery(reader, "2015-10-12", "2016-04-03",
+                false, false, null, null));
+        assertEquals(Relation.INTERSECTS, ft.isFieldWithinQuery(reader, "2015-10-12", "2016-04-03",
+                false, true, null, null));
+        assertEquals(Relation.INTERSECTS, ft.isFieldWithinQuery(reader, "2015-10-12", "2016-04-03",
+                true, false, null, null));
+        IOUtils.close(reader, w, dir);
+    }
```
Comment (colings86) [inline core/src/test/java/org/elasticsearch/index/mapper/core/DateFieldTypeTests.java]: Should we add tests which provide a custom timezone and a custom date math parser (custom date format)?

File: core/src/test/java/org/elasticsearch/index/query/RangeQueryBuilderTests.java [hunk 3]
```diff
@@ -399,317 +397,13 @@ public void testNamedQueryParsing() throws IOException {
...
-        query.from(queryFromValue);
-        query.to(queryToValue);
-        QueryShardContext queryShardContext = queryShardContext();
-        FieldStatsProvider fieldStatsProvider = new FieldStatsProvider(null, null) {
-
-            @Override
-            public Relation isFieldWithinQuery(String fieldName, Object from, Object to, boolean includeLower, boolean includeUpper,
-                    DateTimeZone timeZone, DateMathParser dateMathParser) throws IOException {
+        RangeQueryBuilder query = new RangeQueryBuilder(fieldName) {
+            protected MappedFieldType.Relation getRelation(QueryRewriteContext queryRewriteContext) throws IOException {
```
Comment (colings86) [inline core/src/test/java/org/elasticsearch/index/query/RangeQueryBuilderTests.java]: I'm not a fan of overriding the class under test, but short of mocking IndexReader and MapperService I'm not sure how we can test this here.

Comment (colings86) [inline core/src/test/java/org/elasticsearch/index/query/RangeQueryBuilderTests.java]: The other problem with doing this is that we can't test the logic in `getRelation(QueryRewriteContext) here.

**Comment by colings86** [comment]
@jpountz I left some comments on the tests but I like how much this change simplifies the logic

**Comment by jpountz** [comment]
@colings86 I pushed more commits to improve the testing. Would you mind having another look?

**Comment by colings86** [comment]
@jpountz Thanks for making the changes to the tests. LGTM

**Pull request merged**